packages/core/versioning/core-version-http-server/src/version_http_toolkit.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  IRouter,
+  RouteConfig,
+  RouteMethod,
+  RequestHandler,
+  RouteValidatorFullConfig,
+  RequestHandlerContextBase,
+} from '@kbn/core-http-server';
+
+type RqCtx = RequestHandlerContextBase;
+
+/** A set of type literals to determine accepted versions */
+export type Version = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10';
+
+/** Arguments to create a {@link VersionedRouter | versioned router}. */
+export interface CreateVersionedRouterArgs<Ctx extends RqCtx = RqCtx> {
+  /** A router instance */
+  router: IRouter<Ctx>;
+}
+
+/**
+ * This interface is the starting point for creating versioned routers and routes
+ *
+ * @example
+ * const versionedRouter = vtk.createVersionedRouter({ router });
+ *
+ * ```ts
+ * const versionedRoute = versionedRouter
+ *  .post({
+ *    path: '/api/my-app/foo',
+ *    options: { timeout: { payload: 60000 } },
+ *  })
+ *  // First version of the API, accepts { foo: string } in the body
+ *  .addVersion(
+ *    { version: '1', validate: { body: schema.object({ foo: schema.string() }) } },
+ *    async (ctx, req, res) => {
+ *      await ctx.fooService.create(req.body.foo);
+ *      return res.ok({ body: { foo: req.body.foo } });
+ *    }
+ *  )
+ *  // Second version of the API, accepts { fooName: string } in the body
+ *  .addVersion(
+ *    { version: '2', validate: { body: schema.object({ fooName: schema.string() }) } },
+ *    async (ctx, req, res) => {
+ *      await ctx.fooService.create(req.body.fooName);
+ *      return res.ok({ body: { fooName: req.body.fooName } });
+ *    }
+ *  );
+ * ```
```

Comment:
nit: one of the cons of structuring an API by path instead of by version (our previous discussion) is that it becomes more awkward to make a breaking change to a path. So you have to create a new versionedRouter and it's no longer clear that it shares a "version history" with other paths/controllers.

So I wondered if we should allow a path to be specified in a version too. E.g. changing `'/api/my-app/foo'` -> `'/api/my-app/foobar'`

 ```ts
const versionedRoute = versionedRouter
.post({
   path: '/api/my-app/foo',
   options: { timeout: { payload: 60000 } },
 })
 // First version of the API, accepts { foo: string } in the body
.addVersion(
   { version: '1', validate: { body: schema.object({ foo: schema.string() }) } },
   async (ctx, req, res) => {
     await ctx.fooService.create(req.body.foo);
     return res.ok({ body: { foo: req.body.foo } });
   }
 )
 // Second version of the API, accepts { fooName: string } in the body and new path
.addVersion(
   { 
     version: '2',
     path: '/api/my-app/foobar', // <-- Version 2 makes a breaking change to the path
     validate: { body: schema.object({ fooName: schema.string() }) }
   },
   async (ctx, req, res) => {
     await ctx.fooService.create(req.body.fooName);
     return res.ok({ body: { fooName: req.body.fooName } });
   }
 );
```

This should not be a very common use case but it does happen.

Reply:
That's a great point. I think it could definitely happen with parameters added/removed/changed in the path!

Reply:
I'll incoroporate your changes, but leave this comment open for others!

Reply:
A "_breaking change to a path_" is basicaly a new path so not sure it is a breaking change for the client. (`'/api/my-app/foobar'` is a **new** "v1" route). The example of JL(in `example.ts`) where he changes the parameter name from "name" to "id" is maybe an exception.

Reply:
Good point @sebelga , I realise that example did not capture the nature of the issue correctly. That would be a change that is not visible to consumers of the API. I'll update to capture it correctly.

Reply:
The breaking change in the example is in the body where we change `foo` to `fooName`. I was just hoping to show that you can change name of params if you wanted to.

Reply:
> '/api/my-app/foobar' is a new "v1" route

It could be modeled that way but doesn't have to be. In e.g. alerting we changed our terminology so what was previously called alerts now became rules and actions became alerts. To capture this change in domain language they might want to release a new version of the API where many paths would be different. Having two completely different endpoints both use `alerts` with a different meaning would be confusing so here the new version helps users navigate the change.